server.go

```diff
@@ -352,6 +363,18 @@ func WithoutLogger() func(*Server) {
 	}
 }
 
+// WithListener configures the server to use a custom listener.
+// If no listener is provided, it creates a default listener using the server's address.
+func WithListener(listener net.Listener) func(*Server) {
+	return func(s *Server) {
+		if s.listener != nil {
+			panic("a listener is already configured; cannot overwrite it")
+		}
+		s.listener = listener
+		s.Server.Addr = listener.Addr().String()
+	}
+}
```

Comment:
Please check, but I think there is a need for something like this


```suggestion
func WithListener(listener net.Listener) func(*Server) {
	return func(s *Server) {
		if s.listener != nil {
			panic("a listener is already configured; cannot overwrite it")
		}
		s.listener = listener
		if _, ok := listener.(tls.Listener) {
			t.isTLS = true
		}
		s.Server.Addr = listener.Addr().String()
	}
}
```

This is pseudo code made on my phone, you may need to adapt/fix it

Reply:
I used reflection to check if the instance is of type tls.listener `// WithListener configures the server to use a custom listener.
func WithListener(listener net.Listener) func(*Server) {
	return func(s *Server) {
		if s.listener != nil {
			panic("a listener is already configured; cannot overwrite it")
		}
		s.isTLS = isTLSListener(listener)
		WithAddr(listener.Addr().String())(s)
		s.listener = listener
	}
}

func isTLSListener(listener net.Listener) bool {
	listenerType := reflect.TypeOf(listener)
	if listenerType != nil && listenerType.String() == "*tls.listener" {
		return true
	}
	return false
}
`

Reply:
The code is way better, but using assertion seems overkill, also it's very slow.